2016/proposals/narayan-aditya-visualizer.md

```diff
@@ -0,0 +1,244 @@
+ * Backend  : [Django Rest Framework]
+ * Scheduler: [Cron]
+ * Database : [PostgreSQL]
+ * Frontend : [Charts.js] and [Bootstrap]
+
+### Backend - Django Application
+
+The survey visualizer application would be made using
+[Django Rest Framework] (henceforth called DRF) over [Django]
+to allow easier integration with [AMY] when required.
+
+The tentative API structure exposed by DRF is:
+ * GET `/api/workshop/`  
+   Lists all workshops that have surveys collecting responses.
+ * GET `/api/workshop/<workshop_id>/`  
+   Returns details of a particular workshop.
+ * GET `/api/workshop/<workshop_id>/survey/`  
+   Lists all the surveys whose responses are being collected.
+   Each survey has its corresponding `survey_id`.
+ * POST `/api/workshop/<workshop_id>/survey/refetch`  
+   Repopulates the list of surveys whose response are being
+   collected.
+ * GET `/api/workshop/<workshop_id>/survey/<survey_id>`  
+   Returns the responses for the particular survey, listed
+   according to the questions contained in the survey.
+ * POST `/api/workshop/<workshop_id>/survey/<survey_id>/refetch`  
+   Instructs the Django application to fetch the responses
+   that the survey has received since the last time the
+   SurveyMonkey API was hit.
+
+The underlying Django app will perform these tasks:
+ * Polling SurveyMonkey API management command:
+   This management command would populate the database with survey
+   responses fetched from the SurveyMonkey API.
+   The number of survey responses obtained from the [get_response_counts]
+   API call will be compared to the number of survey responses
+   saved in the database. The remaining responses will be fetched
+   using the [get_respondent_list] and [get_responses] API calls.
+ * Segregating the survey responses according to the answer to the
+   question: *Which workshop did you attend?* or the custom variable
+   `workshop_id`.
+ * Rendering and serving the templates.
+
+### Scheduler - Cron
+
+[Cron] will be used as a job scheduler to periodically run
+the *Polling SM API* management command at fixed time intervals.
+The interval duration would be dynamically determined by the
+number of API calls left and the time of the day.
+
+*Why Cron?*
+ * Cron is the easiest job scheduling solution available to us
+   and is available across all Unix environments.
+ * > Premature optimization is the root of all evil
+
+   -- DonaldKnuth
+
+### Database - PostgreSQL
+
+The Django app will use [PostgreSQL] as its database engine to
+store the survey responses fetched by the backend.
+
+*Why PostgreSQL?*
+ *  Django provides [PostgreSQL-specific features] like the ability
+    and query JSONs, dicts and arrays. The database schema, when
+    finalized, would benefit from these features.
+ *  PostgreSQL is actively maintained and is seen to perform
+    faster writes as compared to SQLite.
+ *  Although AMY currently uses SQLite database, a [migration]
+    is currently being planned to upgrade AMY to PostgreSQL.
```

Comment:
> As for the response content you suggested different tables… Maybe you could store JSON in there?

I would suggest that we parse the JSONs into SQL columns.

While storing JSONs as strings in a text field, we may have to deal with:
* Parsing: We will have to parse it on every database hit.
* Validation: We will have to implement unique constraints ourselves.
* Maintaining the integrity across relationships: Manually delete a row in one table when the reference from another table is deleted.
* Querying: How do we run a query across JSONs that are stored as text?
Maybe there exists solutions that would take care of the above problems on SQLite too.

While storing them in tables, we have to use a [GenericForeignKey](https://docs.djangoproject.com/en/1.9/ref/contrib/contenttypes/#django.contrib.contenttypes.fields.GenericForeignKey) that does not support queries across it.
But we aren't interested in such queries.
We would rather query
> give me all answers to the question  *Please select the workshop you attended?*

over
> give me responses that answer the question *Please select the workshop you attended?*

The problems we may face while storing JSONs as text outweigh the debt of using a polymorphic relationship from a Response ---> Question.